src/libponyrt/sched/scheduler.c

```diff
@@ -470,7 +575,8 @@ static bool quiescent(scheduler_t* sched, uint64_t tsc, uint64_t tsc2)
 
   // only scheduler 0 can initiate shutdown (it is the ony that gets all the
   // ACK messages as part of the CNF/ACK coordination for shutdown)
-  if(0 == sched->index)
+  // only if there are no pinned actors in the system...
+  if(0 == sched->index && 0 == get_pinned_actor_count())
```

Comment:
why the check against the pinned actor count for quiescence?

Reply:
so an app with only pinned actors doesn't shut down too early..

Reply:
pinning doesn't prevent garbage collection of the actor so, i don't think this is needed.

i think what is needed is the usual "there's a noisy" or similiar.

we discussed this a good deal on the sync call. let's find some time to have a synchronous conversation, you and i about this.

Reply:
> pinning doesn't prevent garbage collection of the actor so, i don't think this is needed.

i don't see how garbage collection of pinned actors is related to this..

> i think what is needed is the usual "there's a noisy" or similiar.

sure, that's another implementation option.. the current option falls into `or similar` in my opinion because i expect the atomic variable used have little to no contention so that's what i implemented..

> we discussed this a good deal on the sync call. let's find some time to have a synchronous conversation, you and i about this.

yep.. sounds like a plan..

Reply:
why would an app with only pinned actors "shutdown early"?

Reply:
> why would an app with only pinned actors "shutdown early"?

because, in the current implementation, the `pinned actor scheduler thread` doesn't participate in the `CNF`/`ACK` protocol that all the regular scheduler threads use to determine `quiescence`..

Reply:
Ah. Hmmm...

what's the overhead if it did participate?

Reply:
not sure.. i didn't think through that implementation option.. probably not huge, but it would add code complexity in that all of the `quiescence`/`CNF`/`ACK` logic is tied to work stealing (and all the other logic around work stealing like blocking/unblocking/suspending) which the `pinned actor scheduler thread` also does not participate in currently and we'd have to figure out how to reconcile all that..

Reply:
yeah.

i the concern with the current approach is that pinning an actor changes the semantics of shutdown etc. And there's weird edge cases like...

what happens if an actor is gc'd before it has a chance to unpin itself. if an external actor gives up a reference and the pinned actor isn't noisy, then it can be reaped without it doing anything. And then, no shutdown.

I think we should aim for no change in terms of shutdown for pinned actors vs other actors. I'm not sure what the right balancing of things is, but that's my general feeling modulo literally everything I haven't thought about or considered yet.

Thoughts on how to accomplish in a way that is perhaps not being part of CNF/ACK et al?

Reply:
> i the concern with the current approach is that pinning an actor changes the semantics of shutdown etc.

yes and no.. in theory, as long as the cycle detector is `perfect` and `guaranteed to reap any quiescent cycles/actors`, the shutdown behavior semantics are not changed at all as long as the cycle detector is enabled.. but yes, the semantics are changed when the cycle detector is disabled...

> And there's weird edge cases like...
> what happens if an actor is gc'd before it has a chance to unpin itself. if an external actor gives up a reference and the pinned actor isn't noisy, then it can be reaped without it doing anything. And then, no shutdown.

no, if an actor is reaped/gc'd and destroyed via `ponyint_actor_destroy` then the `pinned_actor_count` is decremented automagically per the logic changes to `ponyint_actor_destroy` included in this PR and yes, shutdown..

> Thoughts on how to accomplish in a way that is perhaps not being part of CNF/ACK et al?

the CNF/ACK process is how the various schedulers coordinate that they all have no actors to run and that quiescence has been reached.. not sure if it'd be easier to do something separate from CNF/ACK for the pinned actor thread or to have it participate in it somehow.. it's something that would need some thought..

> I think we should aim for no change in terms of shutdown for pinned actors vs other actors. I'm not sure what the right balancing of things is, but that's my general feeling modulo literally everything I haven't thought about or considered yet.

i go back to my earlier comment about maybe marking this functionality as `experimental` for now and merging as is and then iterating on the quiescence logic after merge vs trying to get everything perfect as part of this PR..

Reply:
so if noblock is on and there are no references to the actor then it will be reaped and then quiescence can be achieved.

if noblock is on and there's a cycle, quiescence won't be reached regardless of if the actors are pinned or not. 

what is the scenario where not unpinning results in a lack of quiescence that would arise from a scenario that would already prevent quiescence?

Reply:
> so if noblock is on and there are no references to the actor then it will be reaped and then quiescence can be achieved.
> 
> if noblock is on and there's a cycle, quiescence won't be reached regardless of if the actors are pinned or not.

if you say so.. i haven't looked at that code/thought about the exact behavior of `noblock` in a long time so i am not able to confirm your statement..

> what is the scenario where not unpinning results in a lack of quiescence that would arise from a scenario that would already prevent quiescence?

i can't answer that question because i haven't looked at the code/thought about the exact behavior of `noblock` in a long time and i am also not sure that the `cycle detector` is actually `perfect` and `guaranteed to reap any quiescent cycles/actors` so i can't be sure there aren't any scenarios where the cycle detector might not reap cycles that include a `pinned actor` that could in theory be reaped which would prevent `quiescence` due to the `pinned actor` but wouldn't if the same unreaped cycle did not include any `pinned actors`..

Reply:
i think we can remove the unpinning caveat from the documentation. i don't believe it applies. i think what you have done is correct and should work with the cycle detector and noblock regardless of pinned/not pinned.

i think we can roll this out as you say as "experimental" with a note in the release notes that it is possible that this could cause runtime hangs but we don't believe it will.

Reply:
> if noblock is on and there's a cycle, quiescence won't be reached regardless of if the actors are pinned or not.

@SeanTAllen i just checked and currently `noblock` doesn't prevent quiescence even if there are cycles..

i built `examples/ring` and ran it with both
```
./ring --size 5 --pass 5 --count 5
```
and
```
./ring --size 5 --pass 5 --count 5 --ponynoblock
```
and both ran and reached quiescence and exited..

the change with this PR is that with `pinned actors`, the one with `--ponynoblock` would no longer exit unless the programmer manually `unpins` the all `pinned actors` first..

Reply:
I think that is problematic. I need refresh my knowledge of the quiescence logic and why the noblock and cycle would exit.

Reply:
to be clear, you changed ring to pin at least 1 actor for the test you mention above, correct?

Reply:
> to be clear, you changed ring to pin at least 1 actor for the test you mention above, correct?

no.. my test was without any pinned actors.. your statement indicated that `noblock would prevent quiescent if there are any cycles even with no pinned actors` (or i misunderstood your statement to imply this) and i wanted to confirm that behavior...

> I think that is problematic. I need refresh my knowledge of the quiescence logic and why the noblock and cycle would exit.

what part do you think is problematic? the existing behavior or the change in behavior with this PR and `pinned actors` or both?

also, my understanding of the `cycle detector` (and by extension `noblock`) is that the `cycle detector` helps prevent memory leaks by reaping cycles of dead actors but it has no impact on `quiescence` since `quiescence` is about whether any actors in the system have any pending messages to process or not (and that has no relation to whether any particular actor is in a cycle or not)..

Reply:
> what part do you think is problematic? the existing behavior or the change in behavior with this PR and pinned actors or both?

I think the ideal would be for you to be able to pin actors and have everything "continue to work as before". Whether this is doable with the various trade-offs is what I want to look at.

Reply:
> no.. my test was without any pinned actors.. your statement indicated that noblock would prevent quiescent if there are any cycles even with no pinned actors (or i misunderstood your statement to imply this) and i wanted to confirm that behavior...

I was incorrect on that front.

Reply:
> no.. my test was without any pinned actors.. 

So this change, without any actors pinned causes the ring example to not exit if noblock is used?

Reply:
> > no.. my test was without any pinned actors..
> 
> So this change, without any actors pinned causes the ring example to not exit if noblock is used?

no, this change doesn't impact pre-existing behavior is no actors are pinned.. the `quiescence` logic is impacted only if actors are pinned...

Reply:
> > what part do you think is problematic? the existing behavior or the change in behavior with this PR and pinned actors or both?
> 
> I think the ideal would be for you to be able to pin actors and have everything "continue to work as before". Whether this is doable with the various trade-offs is what I want to look at.

yes, we discussed this already in comments yesterday and landed on `experimental` for now while we iterate on the `quiescence` implementation details in future PRs as opposed to trying to be perfect with this PR..

Reply:
> > > what part do you think is problematic? the existing behavior or the change in behavior with this PR and pinned actors or both?
> > 
> > 
> > I think the ideal would be for you to be able to pin actors and have everything "continue to work as before". Whether this is doable with the various trade-offs is what I want to look at.
> 
> yes, we discussed this already in comments yesterday and landed on `experimental` for now while we iterate on the `quiescence` implementation details in future PRs as opposed to trying to be perfect with this PR..

yup. i'm ok with that approach. i am trying to understand better to craft what i think would be a more clear "caveat" message.